### PR TITLE
Third Design Review for new DFW API 

### PR DESCRIPTION
**Pull request by madhephaestus**
This update makes the button functions a flat one-liner. The comments on the competion has beed re-written. The begin function does not take a parameter since it can only be 9600.

### CODE REVIEW

File: DFW.cpp [hunk 1]
```diff
@@ -10,298 +10,282 @@
...
-			byteAn[4] = inByteA[4];
-			byteAn[5] = inByteA[5];
-			digitalWrite(debuginpin, 0);
+		break;
+	case waitForAuto:
+		//Serial.println("waiting for auto start");
+		if (start()) {
+			state = Autonomous;
+			autoStartTime = millis(); // sets start time of autonomous
+			// fall through when a state changes
```
Comment (AustinShalit) [inline DFW.cpp]: Do we want to fall through or wait for the next loop?

File: DFW.cpp [hunk 2]
```diff
@@ -10,298 +10,282 @@
...
-	if (Serial1.find('B')) {
-		for (int i = 0; i < 3; i++) {
-			inByteB[i] = Serial1.read();
+		break;
+
+	case Autonomous:
+		timeDiff = millis() - autoStartTime;
+		if (timeDiff > autoTime) {
+			state = waitForTeleop;
+			// fall through when a state changes
```
Comment (AustinShalit) [inline DFW.cpp]: Do we want to fall through or wait for the next loop?

File: DFW.cpp [hunk 3]
```diff
@@ -10,298 +10,282 @@
...
-			//   Serial.println("B Lock");
-			for (int i = 0; i < 5; i++) {
-				byteBu[i] = inByteB[i];
+		break;
+	case waitForTeleop:
+			//Serial.println("waiting for teleop start");
+			if (start()) {
+				state = Teleop;
+				teleopStartTime = millis(); // sets start time of autonomous
+				// fall through when a state changes
```
Comment (AustinShalit) [inline DFW.cpp]: Do we want to fall through or wait for the next loop?

File: DFW.cpp [hunk 4]
```diff
@@ -10,298 +10,282 @@
...
+
+}
+void DFW::update(void) {
+	while (Serial1.peek() != 'A') {
+		char tmp = Serial1.read();
 	}
+	if (Serial1.peek() == 'A' && Serial1.available() >= packetSize) {
+		packet[0] = Serial1.read();
+		if (Serial1.peek() == 'c' && Serial1.available() >= (packetSize - 1)) {
+			//Serial.print("\r\nPacket start = ");
```
Comment (AustinShalit) [inline DFW.cpp]: We should probably remove this comment

File: DFW.cpp [hunk 5]
```diff
@@ -10,298 +10,282 @@
...
+			for (int i = 1; i < packetSize; i++) {
+				packet[i] = Serial1.read();
 
-	while (Serial1.available() > 0) {
-		char clearing = Serial1.read();
+			}
+			if (	packet[0] == 'A' &&
```
Comment (AustinShalit) [inline DFW.cpp]: Is there an extra tab on this line?

File: DFW.cpp [hunk 6]
```diff
@@ -10,298 +10,282 @@
...
+				 Serial.print("(\\n)");
+				 }else
+				 if(packet[i]=='\r'){
+				 Serial.print("(\\r)");
+				 }else{
+				 Serial.print("  ");
+				 Serial.print(packet[i], HEX);
+				 }
+				 }
+				 */
```
Comment (AustinShalit) [inline DFW.cpp]: We should probably remove this comment.

File: DFW.cpp [hunk 7]
```diff
@@ -10,298 +10,282 @@
...
 		digitalWrite(debuginpin, 1);
 		for (int i = 0; i < 2; i++) {
 			byteBu[i] = 127;
 		}
-		for (int i = 0; i <= 4; i++) {
+		for (int i = 0; i < 5; i++) {
 			byteAn[i] = 90;
 		}
+
+	}else{
```
Comment (AustinShalit) [inline DFW.cpp]: Whitespace around else

File: DFW.cpp [hunk 8]
```diff
@@ -10,298 +10,282 @@
...
+		long timeSinceFlash = millis() - flashTime;
+		if(timeSinceFlash>1000){
+			flashTime=millis();
+			digitalWrite(debuginpin, 1);
+		}else{
```
Comment (AustinShalit) [inline DFW.cpp]: Whitespace around else

File: DFW.cpp [hunk 9]
```diff
@@ -10,298 +10,282 @@
...
 			byteBu[i] = 127;
 		}
-		for (int i = 0; i <= 4; i++) {
+		for (int i = 0; i < 5; i++) {
 			byteAn[i] = 90;
 		}
+
+	}else{
+		long timeSinceFlash = millis() - flashTime;
+		if(timeSinceFlash>1000){
```
Comment (AustinShalit) [inline DFW.cpp]: Whitespace after `if` and before `{`

File: DFW.cpp [hunk 10]
```diff
@@ -10,298 +10,282 @@
...
 
 DFW::DFW(int debugpin)    //13 is easiest
 		{
 	pinMode(debugpin, OUTPUT);
 	digitalWrite(debugpin, 0);
 	debuginpin = debugpin;
+	myAutonomous = Null;
+	myTeleop = Null;
+	state = powerup;
+	//printing = false;
```
Comment (AustinShalit) [inline DFW.cpp]: Remove this comment

File: DFW.cpp [hunk 11]
```diff
@@ -10,298 +10,282 @@
...
-		break;
-	default:
-		Serial1.begin(baud);
-		break;
-	}
 }
 
 int DFW::joystickrv(void) // returns joystick right vertical
 		{
 //Serial.println("jrv");
```
Comment (AustinShalit) [inline DFW.cpp]: We do not need this comment

File: DFW.cpp [hunk 12]
```diff
@@ -10,298 +10,282 @@
...
-	unsigned char stickRv = byteAn[1];
+	unsigned char stickRv = byteAn[0];
 	return (stickRv);
 }
```
Comment (AustinShalit) [inline DFW.cpp]: Can this function just return `byteAn[0]` instead of copying it to another piece of memory?

File: DFW.cpp [hunk 13]
```diff
@@ -10,298 +10,282 @@
...
+bool DFW::start(void) // returns start button state
 		{
-	int startB;
-	if (((byteBu[1] & 0b00000010) == 0b00000010)) {
-		startB = 1;
-	} else {
-		startB = 0;
-	}
-	return (startB);
+	return ((byteBu[1] & 0b00000010) == 0) ;
```
Comment (AustinShalit) [inline DFW.cpp]: For all of these functions:

- remove space before `;`
- remove extra parenthesis around the expression

File: examples/DFWRBE1001Template/DFWRBE1001Template.ino [hunk 14]
```diff
@@ -1,90 +1,23 @@
...
+ * This Template
+ * is designed to run the autonomous and teleop sections of the final
+ * competition. Write and test your autonomous and teleop code on your
+ * own and place the code in auto.cpp or teleop.cpp respectively.
+ * The functions will be called by the competition framework based on the
+ * time and start button. DO NOT change this file, your code will be called
+ * by the framework. The framework will pass your code a reference to the DFW
+ * object as well as the amount of MS remaining.
+ */
+#include <DFW.h> // DFW include
```
Comment (AustinShalit) [inline examples/DFWRBE1001Template/DFWRBE1001Template.ino]: Is this comment needed?  Seems very obvious.

File: examples/DFWRBE1001Template/auto.cpp [hunk 15]
```diff
@@ -0,0 +1,6 @@
+#include <DFW.h> // DFW include
+
+void autonomous( long time,DFW & dfw) { // function definition
```
Comment (AustinShalit) [inline examples/DFWRBE1001Template/auto.cpp]: Is this comment needed?  Seems very obvious.

File: examples/DFWTank/DFWTank.ino [hunk 16]
```diff
@@ -1,40 +1,31 @@
...
+
+void setup() {
+  Serial.begin(9600); // Serial output begin. Only needed for debug
+  dfw.begin(); // Serial1 output begin for DFW library. Buad and port #."Serial1 only"
+  leftmotor.attach(4, 1000, 2000); // left drive motor pin#, pulse time for 0,pulse time for 180
+  rightmotor.attach(5, 1000, 2000); // right drive motor pin#, pulse time for 0,pulse time for 180
+}
+
+void loop() {
+  dfw.run();// Called to update the controllers output. Do not call faster than every 15ms.
```
Comment (AustinShalit) [inline examples/DFWTank/DFWTank.ino:25]: It can now be called that fast

File: examples/WirelessControllercodeRBE/WirelessControllercodeRBE.ino [hunk 17]
```diff
@@ -1,141 +1,140 @@
...
+  Serial1.write(joy2);//0-180 10
+  byte joy3 = joystick[3];
+  Serial1.write(joy3);//0-180 11
+  Serial1.print('a');// end byte 12
+  Serial1.println(); // 13
+}
+
+
+
+
```
Comment (AustinShalit) [inline examples/WirelessControllercodeRBE/WirelessControllercodeRBE.ino]: Remove extra new lines

**Review by AustinShalit — CHANGES_REQUESTED** [review]

File: examples/DFWTank/DFWTank.ino (hunk 16 above)
Comment (madhephaestus) [reply examples/DFWTank/DFWTank.ino:25]: yes it can

**Pull request merged**